lib/rspec/rails/matchers/send_email.rb

```diff
@@ -0,0 +1,100 @@
+module RSpec
+  module Rails
+    module Matchers
+      # @api private
+      #
+      # Matcher class for `send_email`. Should not be instantiated directly.
+      #
+      # @see RSpec::Rails::Matchers#send_email
+      class SendEmail < RSpec::Rails::Matchers::BaseMatcher
+        def initialize(criteria)
+          normalized = %i[to from].map { |attr| criteria.key?(attr) ? [attr, Array(criteria[attr])] : nil }.compact.to_h
+          @criteria = { **criteria, **normalized }
+        end
+
+        # @private
+        def supports_value_expectations?
+          false
+        end
+
+        # @private
+        def supports_block_expectations?
+          true
+        end
+
+        def matches?(block)
+          define_matched_emails(block)
+
+          @matched_emails.one?
+        end
+
+        private
+
+        def diffable?
+          true
+        end
+
+        def match_when_negated(block)
+          define_matched_emails(block)
+          @matched_emails.empty?
+        end
+
+        def failure_message
+          if multiple_match?
+            'More than 1 matching emails were sent.'
+          else
+            'No matching emails were sent.'
+          end
+        end
+
+        def failure_message_when_negated
+          "Expected not to send an email but it was sent."
+        end
+
+        def deliveries
+          ActionMailer::Base.deliveries
+        end
+
+        def define_matched_emails(block)
+          before = deliveries.clone
+
+          block.call
+
+          after = deliveries
+
+          @diff = after - before
+          @matched_emails = select_matching(@diff)
+        end
+
+        def select_matching(emails)
+          @criteria.empty? ? emails : emails.select(&method(:matched_email?))
+        end
+
+        def matched_email?(email)
+          @criteria.all? do |attr, expected|
+            values_match?(expected, email.public_send(attr))
```

Comment:
Continued from above:
```suggestion
        def matched_email?(email)
          @criteria.all? do |attr, value|
            expected =
              case attr
              # :to / :from are arrays
              when :to,:from then Array(value)
              else
                value
              end

            values_match?(expected, email.public_send(attr))
```